CLDR-19035 GenerateExampleDependencies, performance, refactoring

btangmu opened the pull request:
-Update ExampleDependencies.java by running GenerateExampleDependencies

-Improve GenerateExampleDependencies performance roughly 30% by conditionally skipping intern() in XMLSource: new XMLSource.setDoInternStrings()

-Add options to GenerateExampleDependencies for testing with subset of locales (e.g., -f^[^_]+$ to skip locales whose names contain underscore)

-Report how many minutes it takes to run GenerateExampleDependencies

-Fix a bug in GenerateExampleDependencies: call clearRecordedPaths after loop

-Refactor ICUServiceBuilder and related code: new constructor with CLDRFile, instead of calling setCldrFile separately

-Remove ICUServiceBuilder.setCldrFile and the no-arg ICUServiceBuilder constructor

-Only call the ICUServiceBuilder constructor locally from forLocale, except for fictional test locales xx, mul

-Do not call icuServiceBuilder.clearCache when it is superfluous immediately after constructing

-Call ICUServiceBuilder.clearCache instead of duplicating the code; remove ISB_CAN_CLEAR_CACHE

-New constructors for DateTimeFormats and ZoneFormats obviate calling set()

-New FlexibleDateFromCLDR constructor with CLDRFile replaces FlexibleDateFromCLDR.set

-Change CheckNumbers.defaultNumberingSystem to a local variable

-New constant LocaleNames.XX_TEST (compare existing LocaleNames.MUL)

-Make data private/final/static where possible; ICUServiceBuilder.supplementalData can be static since it does not depend on the locale

-Some automated refactoring by IntelliJ, from accepting suggestions linked to compiler warnings

-Comments

CLDR-19035

- [ ] This PR completes the ticket.

<!--
Thank you for your pull request.
Please see https://cldr.unicode.org/index/process for general
information on contributing to CLDR.

1. Make sure the ticket is filed at
https://unicode-org.atlassian.net/projects/CLDR/
2. Update the PR title and first line of this
message to include the ticket ID (CLDR-_____)
3. You will be automatically asked to sign the contributors’
license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/cldr
- license: https://www.unicode.org/copyright.html#License
-->

ALLOW_MANY_COMMITS=true

btangmu committed:
CLDR-19035 GenerateExampleDependencies, performance, refactoring

-Update ExampleDependencies.java by running GenerateExampleDependencies

-Improve GenerateExampleDependencies performance roughly 30% by conditionally skipping intern() in XMLSource: new XMLSource.setDoInternStrings()

-Add options to GenerateExampleDependencies for testing with subset of locales (e.g., -f^[^_]+$ to skip locales whose names contain underscore)

-Report how many minutes it takes to run GenerateExampleDependencies

-Fix a bug in GenerateExampleDependencies: call clearRecordedPaths after loop

-Refactor ICUServiceBuilder and related code: new constructor with CLDRFile, instead of always calling setCldrFile separately

-Do not call icuServiceBuilder.clearCache when it is superfluous immediately after constructing

-Call ICUServiceBuilder.clearCache instead of duplicating the code; remove ISB_CAN_CLEAR_CACHE

-New constructors for DateTimeFormats and ZoneFormats obviate calling set()

-Change data/methods to private and/or final where possible

-Some automated refactoring by IntelliJ, from accepting suggestions linked to compiler warnings

-Comments

btangmu commented on tools/cldr-code/src/main/java/org/unicode/cldr/test/ExampleDependencies.java:
```diff
@@ -487,9 +498,6 @@ public class ExampleDependencies {
                             "//ldml/personNames/personName[@order=\"([^\"]*+)\"][@length=\"([^\"]*+)\"][@usage=\"([^\"]*+)\"][@formality=\"([^\"]*+)\"]/namePattern")
-                    .putAll(
-                            "//ldml/dates/timeZoneNames/zone[@type=\"([^\"]*+)\"]/short/generic",
-                            "//ldml/personNames/personName[@order=\"([^\"]*+)\"][@length=\"([^\"]*+)\"][@usage=\"([^\"]*+)\"][@formality=\"([^\"]*+)\"]/namePattern")
```
This is the only example dependency removed. I don't know why. The implication is that these namePattern paths no longer depend on these timeZoneNames paths.

btangmu committed:
CLDR-19035 GenerateExampleDependencies: remove experimental rootOnlyPaths

-The idea of rootOnlyPaths is that if a path only occurs in root.xml, and root is read-only as far as ST is concerned, then it does not matter if that examples for other paths depend on that path value, since the value will not change. But, that is untrue, since a non-root locale could override the root value if ST allows. The idea might be worth further investigation in the case of paths that really cannot be changed in ST.

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/ICUServiceBuilder.java:
```diff
@@ -40,6 +40,18 @@ public class ICUServiceBuilder {
     }
 
+    public ICUServiceBuilder() {
+        // This constructor without CLDRFile requires setCldrFile to be called subsequently.
```
Is the idea to make ICUServiceBuilder be immutable, thus no set() method. (Seems like a good idea).

Also, splitting out the functionality of ICUServiceBuilder into separate classes with smaller chunks of data that they need would allow us to be finer-grained when flushing a cache.

btangmu replied:
The fourth commit replaces nearly all `new ICUServiceBuilder(cldrFile)` with `ICUServiceBuilder.forLocale()` for consistency and improved performance. The only exception is for some unit tests that use fictitious locale names "xx" and "mul", with special test data, for which `forLocale` would not work.

There is a performance improvement for `TestCheckCLDR.TestAllLocales` which was taking between 91 and 107 seconds locally and now takes between 75 and 79 seconds.

btangmu replied:
Before 4th commit, "Test with maven" took 22m 28s; after 4th commit, "Test with maven" took 21m 25s. Only 63 seconds shorter, so not a big performance improvement as far as running tests is concerned. The code is cleaner though, and these changes may pave the way for more improvements to GenerateExampleDependencies and other features.

macchiati replied:
> It could be helpful to make ICUServiceBuilder immutable. I'm still trying to understand it better, especially in relation to caching and example dependencies.

This is a good topic for discussion in the infra group.

When the underlying CLDRFile is immutable (or frozen), then the ICUServiceBuilder can be cached. So for charts, unit tests, etc that can be the case. We should check, however, that CLDRFile::isFrozen! (In general, we should not be caching mutable/unfrozen objects.

The problem is when the CLDRFile can change underneath, as in the survey tool. In that case, it cannot be cached, since the change to CLDRFile can require that the data in the ICUServiceBuilder be refreshed. Now, ideally, we would have objects like ICUServiceBuilder have Listeners on their mutable CLDRFile; then we would know to refresh the ICUServiceBuilder only when that CLDRFile changes a value that affects the ICUServiceBuilder; that would cut down on needless recreation. (But before undertaking anything like that, we should measure the performance cost of needless recreation first!)

The other nice cleanup would be to split ICUServiceBuilder into different parts, based on the CLDRData that they need. For example, we could have ICUNumberBuilder that just fetched the number information and ICUDateBuilder that just fetched the date information (but also has-a ICUNumberBuilder, since that is used in building dates). But again, only worth doing if we know it would improve performance significantly.

> Only 63 seconds shorter

That is actually pretty good! That is 5% of 22m 28s. With ten 63s reductions, we'd be taking half the time for the tests (and probably a measurable effect on ST; since checks and examples would run faster).

srl295 replied:
> The fourth commit replaces nearly all `new ICUServiceBuilder(cldrFile)` with `ICUServiceBuilder.forLocale()` for consistency and improved performance. The only exception is for some unit tests that use fictitious locale names "xx" and "mul", with special test data, for which `forLocale` would not work.
> 
> There is a performance improvement for `TestCheckCLDR.TestAllLocales` which was taking between 91 and 107 seconds locally and now takes between 75 and 79 seconds.

It's also causing CLDR-19409 because some clients should be using SurveyTool's factory instead of Factory.make

btangmu committed:
CLDR-19035 GenerateExampleDependencies: more ICUServiceBuilder-related refactoring

-Remove the no-arg ICUServiceBuilder constructor

-Remove ICUServiceBuilder.setCldrFile

-New FlexibleDateFromCLDR constructor with CLDRFile replaces FlexibleDateFromCLDR.set

-Change CheckNumbers.defaultNumberingSystem to a local variable

-Make data private and/or final where possible

btangmu committed:
CLDR-19035 Replace ICUServiceBuilder constructor with forLocale

-Only call the ICUServiceBuilder locally from forLocale, except for fictional test locales xx, mul

-New constants LocaleNames.XX_TEST (compare existing LocaleNames.MUL)

-Make data private/final/static where possible; ICUServiceBuilder.supplementalData can be static since it does not depend on the locale

srl295 commented on tools/cldr-code/src/main/java/org/unicode/cldr/icu/TestCldr.java:
```diff
@@ -223,7 +224,7 @@ void set(AttributeName attributeName, String attributeValue) {
 
         void checkResult(String value) {
-            ULocale ul = new ULocale("xx");
+            ULocale ul = new ULocale(LocaleNames.XX_TEST);
```
why XX and not `ZZ`?

btangmu replied:
I have no idea! I don't like to see a mysterious locale ID like "xx" hidden away as a string; defining a constant in LocaleNames.java is a step towards clarifying what it is. If it can be removed, so much the better. I can address that in a follow-up PR.

srl295 commented on tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestDateOrder.java:
```diff
@@ -46,7 +47,7 @@ public void TestDateImportance() {
             // Build test file
 
-            XMLSource source = new SimpleXMLSource("xx");
+            XMLSource source = new SimpleXMLSource(LocaleNames.XX_TEST);
```
```suggestion
            XMLSource source = new SimpleXMLSource("mul");
```

mul seems good for here

btangmu replied:
Again, let's consider that (with `LocaleNames.MUL` not a string literal) for a follow-up -- such a change is somewhat out-of-scope for this ticket, which is about example dependencies...

srl295 commented on tools/cldr-code/src/main/java/org/unicode/cldr/util/ICUServiceBuilder.java:
```diff
@@ -40,6 +40,40 @@ public class ICUServiceBuilder {
+    /**
+     * This constructor is meant to be called only by the private constructor (with locIsFake =
+     * false) and by unit tests with fictional locale ID such as "xx" or "mul" (with locIsFake =
+     * true) for which ICUServiceBuilder.forLocale won't work
```
```suggestion
     * false) and by unit tests with fictional locale ID such as "mul" (with locIsFake =
     * true) for which ICUServiceBuilder.forLocale won't work
```

drop reference to xx, don't use it.

srl295 reviewed (approved):
approving but we should drop references to locale id "xx" - no reason for it.

also 'mul' is the valid sandbox locale.

macchiati commented on tools/cldr-code/src/main/java/org/unicode/cldr/icu/TestCldr.java:
```diff
@@ -223,7 +224,7 @@ void set(AttributeName attributeName, String attributeValue) {
 
         void checkResult(String value) {
-            ULocale ul = new ULocale("xx");
+            ULocale ul = new ULocale(LocaleNames.XX_TEST);
```
I wouldn't remove anything without knowing what it is doing; ULocale("xx") might just be a locale made to never overlap with a real CLDR locale.
